tutorial/lesson_19_staging_func_or_image_param.cpp

```diff
@@ -0,0 +1,289 @@
+                    return -1;
+                }
+            }
+        }
+    }
+
+    {
+        // Func::in() can also be used to group multiple stages of a Func
+        // in the same loopness. Consider the following code:
+        Func f("f"), g("g");
+        f(x, y) = x + y;
+        f(x, y) += x - y;
+        g(x, y) = x * y * f(x, y);
+
+        // When we schedule 'f' to be computed at root (by calling f.compute_root()),
+        // all its stages are computed at separate loopness:
+        // for y:
+        //   for x:
+        //     f(x, y) = x + y
+        // for y:
+        //   for x:
+        //     f(x, y) += x - y
+        // for y:
+        //   for x:
+        //     g(x, y) = x * y * f(x, y)
+
+        // We can use Func::in() to group those stages to be computed at the
+        // same loopness like so:
+        f.in(g).compute_root();
+
+        Buffer<int> halide_result = g.realize(20, 20);
+
+        // The equivalent C is:
+        int f_wrapper[20][20];
+        for (int y = 0; y < 20; y++) {
+            for (int x = 0; x < 20; x++) {
+                int f = x + y;
+                f += x - y;
+                f_wrapper[y][x] = f;
+            }
+        }
+        int c_result_g[20][20];
+        for (int y = 0; y < 20; y++) {
+            for (int x = 0; x < 20; x++) {
+                c_result_g[y][x] = x * y * f_wrapper[y][x];
+            }
+        }
+
+        // Check the results match:
+        for (int y = 0; y < 20; y++) {
+            for (int x = 0; x < 20; x++) {
+                if (halide_result(x, y) != c_result_g[y][x]) {
+                    printf("halide_result(%d, %d) = %d instead of %d\n",
+                           x, y, halide_result(x, y), c_result_g[y][x]);
+                    return -1;
+                }
+            }
+        }
+    }
+
+    {
+        // ImageParam::in() behaves the same way as Func::in(). We can also
+        // use ImageParam::in() to stage loads from an ImageParam via some
+        // intermediate buffer (e.g. on the stack or in shared GPU memory).
+
+        // The following example illustrates how you would use ImageParam::in()
+        // to stage loads from an ImageParam in tiles.
+        ImageParam img(Int(32), 2, "img");
+        Func f("f");
+        f(x, y) = 2 * img(y, x);
+
+        // First, we tile 'f' into 8x8 blocks.
+        Var xi("xi"), yi("yi");
+        f.compute_root().tile(x, y, xi, yi, 8, 8);
+        // Then, we create a wrapper for 'img' which will load the corresponding
+        // values from 'img' at tile level.
+        Func img_wrapper = img.in().compute_at(f, x);
+        // If, for some reason, we want to unroll the first dimension of the
+        // wrapper by 2, we can do the following:
+        img_wrapper.unroll(_0, 2);
+        // Note that we use implicit variables to name the dimensions of the image
+        // wrapper: _0 as the 1st dimension, _1 as the 2nd dimension, and so on.
```

Comment:
Say *why* we do this: because the anonymous wrapper Func created by `in` does not have any explicitly named variables. (This is presumably specific to ImageParam::in, and not Func::in.)